dbms/src/Storages/MergeTree/KeyCondition.cpp

```diff
@@ -444,6 +444,45 @@ void KeyCondition::traverseAST(const ASTPtr & node, const Context & context, Blo
                   */
                 if (i != 0 || element.function == RPNElement::FUNCTION_NOT)
                     rpn.emplace_back(std::move(element));
+
+                /* Combine statements like (x >= a) and (x <= b) into x in range [a, b] */
+                if (rpn.size() >= 3 && rpn.back().function == RPNElement::FUNCTION_AND)
+                {
+                    const size_t to_modify = rpn.size() - 3;
+                    const auto rhs = rpn[to_modify + 1];
+                    const auto lhs = rpn[to_modify];
+                    if (
+                            lhs.function == RPNElement::FUNCTION_IN_RANGE &&
+                            rhs.function == RPNElement::FUNCTION_IN_RANGE &&
+                            lhs.key_column == rhs.key_column &&
+                            lhs.monotonic_functions_chain == rhs.monotonic_functions_chain)
+                    {
+                        bool merged = false;
+                        if (!lhs.range.right_bounded && !rhs.range.left_bounded)
```

Comment:
Can we just always use `Range::intersectRange`, or is there some case when this doesn't work?